File: ros2_ws/src/diffdrive/diffdrive/cmd_vel.py
import rclpy
from rclpy.node import Node

from geometry_msgs.msg import Twist
from std_msgs.msg import String

import math

class CmdVelNode(Node):
    def __init__(self):
        super().__init__('cmd_vel_node')

        # Read parameters
        self.declare_parameter('wheel_radius', 0.0335)
        self.declare_parameter('wheel_base', 0.215)
        self.declare_parameter('ticks_per_revolution_l', 2006)
        self.declare_parameter('ticks_per_revolution_r', 1992)
        
        # Hae parametrien arvot
        self.wheel_radius = self.get_parameter('wheel_radius').value
        self.wheel_base = self.get_parameter('wheel_base').value
        self.ticks_per_revolution_l = self.get_parameter('ticks_per_revolution_l').value
        self.ticks_per_revolution_r = self.get_parameter('ticks_per_revolution_r').value

        self.get_logger().info(f'Wheel radius: {self.wheel_radius}')
        self.get_logger().info(f'Distance between wheels: {self.wheel_base}')

        self.cmd_vel_subscriber = self.create_subscription(
            Twist,
            'cmd_vel',
            self.cmd_vel_callback,
            10
        )

        self.mc_publisher = self.create_publisher(
            String,
            'motor_command',
            10
        )

    # Convert m/s to motor controller spd value
    def mps_to_spd_l(self, mps):
        # muunnetan m/s moottoriohjaimen spd arvoksi
        pid_freq = 10  # Moottoriohjain päivittyy 10Hz taajuudella

        # kulmanopeus (rad/s)
        radps = mps / self.wheel_radius

        # RPM
        rpm = radps * (60 / (2 * math.pi))

        spd = rpm * (self.ticks_per_revolution_l / 60 / pid_freq)
        return spd
    
    def mps_to_spd_r(self, mps):
        # muunnetan m/s moottoriohjaimen spd arvoksi
        pid_freq = 10  # Moottoriohjain päivittyy 10Hz taajuudella

        # kulmanopeus (rad/s)
        radps = mps / self.wheel_radius

        # RPM
        rpm = radps * (60 / (2 * math.pi))

        spd = rpm * (self.ticks_per_revolution_r / 60 / pid_freq)
        return spd   

    # def mps_to_spd(self,mps):
    #     # muunnetan m/s moottoriohjaimen spd arvoksi
    #     pid_freq = 10 # Moottoriohjain päivittyy 10Hz taajuudella

    #     # kulmanopeus (rad/s)
    #     radps = mps / self.wheel_radius

    #     # RPM
    #     rpm = radps * ( 60 / ( 2 * math.pi ) )
    
    #     spd = rpm * (self.ticks_per_revolution / 60 / pid_freq)
    #     return spd

    def cmd_vel_callback(self, msg):
        # mps_l = vasen rengas m/s
        # mps_r = oikea rengas m/s
        mps_l = (msg.linear.x - (msg.angular.z * self.wheel_base / 2.0))
        mps_r = -(msg.linear.x + (msg.angular.z * self.wheel_base / 2.0))

        string_msg = String()
        spd_l = self.mps_to_spd_l(mps_l)
        spd_r = self.mps_to_spd_r(mps_r)
        if max(abs(spd_l), abs(spd_r)) > 255:
            scale = 255 / max(abs(spd_l), abs(spd_r))
            spd_l *= scale
            spd_r *= scale
            
        string_msg.data = "SPD;%i;%i;"%(spd_l, spd_r)
        
        self.mc_publisher.publish(string_msg)



def main(args=None):
    rclpy.init(args=args)

    cmdvel_node = CmdVelNode()

    try:
        rclpy.spin(cmdvel_node)
    except KeyboardInterrupt:
        pass
    finally:
        cmdvel_node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == '__main__':
    main()

File: ros2_ws/src/diffdrivec/src/cmd_vel.cpp
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "std_msgs/msg/string.hpp"
#include <cmath>
#include <string>

class CmdVelNode : public rclcpp::Node {
public:
    CmdVelNode() : Node("cmd_vel_node") {
        // Lue parametrit
        this->declare_parameter<double>("wheel_radius", 0.1);
        this->declare_parameter<double>("wheel_base", 0.5);

        // Hae parametrien arvot
        wheel_radius_ = this->get_parameter("wheel_radius").as_double();
        wheel_base_ = this->get_parameter("wheel_base").as_double();

        RCLCPP_INFO(this->get_logger(), "Wheel radius: %f", wheel_radius_);
        RCLCPP_INFO(this->get_logger(), "Wheel base: %f", wheel_base_);

        // Luo tilaajat ja julkaisijat
        cmd_vel_subscriber_ = this->create_subscription<geometry_msgs::msg::Twist>(
            "cmd_vel",
            10,
            std::bind(&CmdVelNode::cmdVelCallback, this, std::placeholders::_1)
        );

        mc_publisher_ = this->create_publisher<std_msgs::msg::String>(
            "motor_command",
            10
        );
    }

private:
    void cmdVelCallback(const geometry_msgs::msg::Twist::SharedPtr msg) {
        // Laske vasemman ja oikean pyörän nopeudet
        double vel_l = +((msg->linear.x + (msg->angular.z * wheel_base_ / 2.0)) / wheel_radius_) * 60 / (2 * M_PI);
        double vel_r = -((msg->linear.x - (msg->angular.z * wheel_base_ / 2.0)) / wheel_radius_) * 60 / (2 * M_PI);

        // Luo ja julkaise motor_command-viesti
        auto string_msg = std_msgs::msg::String();
        string_msg.data = "SPD;" + std::to_string(static_cast<int>(vel_l)) + ";" + std::to_string(static_cast<int>(vel_r)) + ";";
        mc_publisher_->publish(string_msg);
    }

    // Parametrit
    double wheel_radius_;
    double wheel_base_;

    // ROS 2 tilaajat ja julkaisijat
    rclcpp::Subscription<geometry_msgs::msg::Twist>::SharedPtr cmd_vel_subscriber_;
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr mc_publisher_;
};

int main(int argc, char *argv[]) {
    rclcpp::init(argc, argv);

    auto node = std::make_shared<CmdVelNode>();
    rclcpp::spin(node);

    rclcpp::shutdown();
    return 0;
}

